### PR TITLE
User class validation methods 

### PR DESCRIPTION
**Pull request by Bahaaib**
track issue here:
https://github.com/ammaratef45/Attendance/issues/31

### CODE REVIEW

**Comment by 0crat** [comment]
@ammaratef45[/z](https://www.0crat.com/u/ammaratef45) please, pay attention to this pull request

**Comment by 0crat** [comment]
Job [#35](https://github.com/ammaratef45/attendance/issues/35) is now in scope, role is `REV`

**Comment by 0crat** [comment]
@ammaratef45[/z](https://www.0crat.com/u/ammaratef45) everybody who has role `REV` is banned at [#35](https://github.com/ammaratef45/attendance/issues/35); I won't be able to assign anyone automatically; consider assigning someone manually (as in [§19](http://www.zerocracy.com/policy.html#19)), or invite more people (as in [§51](http://www.zerocracy.com/policy.html#51)), or remove the job from the scope (as in [§14](http://www.zerocracy.com/policy.html#14))

File: attendance/lib/BackEnd/user.dart [hunk 1]
```diff
@@ -4,35 +4,62 @@ class User {
...
+    } else {
+      return false;
+    }
+  }
+
   void rename(String newName) {
-    _nativeName = newName;
+    if (isValidName(newName)) {
+      _nativeName = newName;
+    }
```
Comment (ammaratef45) [inline attendance/lib/BackEnd/user.dart]: @Bahaaib else throw an exception

File: attendance/lib/BackEnd/user.dart [hunk 2]
```diff
@@ -4,35 +4,62 @@ class User {
...
+    } else {
+      return true;
+    }
   }
 
   void changePhone(String newPhone) {
-    _phone = newPhone;
+    if (isValidNumber(newPhone)) {
+      _phone = newPhone;
+    }
```
Comment (ammaratef45) [inline attendance/lib/BackEnd/user.dart]: @Bahaaib throw an exception here too

File: attendance/test/userdata_test.dart [hunk 3]
```diff
@@ -41,20 +41,17 @@ void main() {
...
-    /*test('should reject names longer than 20 chars', () {
+    test('should reject names longer than 20 chars', () {
+      testUser.rename('');
       testUser.rename('This is a very long name');
-      bool isFit;
-      testUser.nativeName.length < 21 ? isFit = true : isFit = false;
 
-      expect(isFit, true);
-    });*/
+      expect(testUser.nativeName, isEmpty);
```
Comment (ammaratef45) [inline attendance/test/userdata_test.dart]: @Bahaaib expect the thrown exception
assume you made it like `throw Exception('name is too long!');`
then 
```Dart
try {
    testUser.rename('This is a very long name');
    expect(1 ,2); // to fail the test
} catch(ex) {
    expect(ex.message, 'name is too long!');
}

File: attendance/test/userdata_test.dart [hunk 4]
```diff
@@ -68,70 +65,46 @@ void main() {
...
+    test('should reject numbers start with leading zero ', () {
       testUser.changePhone('0123456789');
-      bool hasLeadingZero;
-      testUser.phone.substring(0, 1) != '0'
-          ? hasLeadingZero = false
-          : hasLeadingZero = true;
 
-      expect(hasLeadingZero, false);
-    });*/
+      expect(testUser.phone, isEmpty);
```
Comment (ammaratef45) [inline attendance/test/userdata_test.dart]: @Bahaaib the same as before

File: attendance/lib/BackEnd/user.dart (hunk 1 above)
Comment (Bahaaib) [reply attendance/lib/BackEnd/user.dart]: Sure

File: attendance/lib/BackEnd/user.dart (hunk 2 above)
Comment (Bahaaib) [reply attendance/lib/BackEnd/user.dart]: Okay

**Comment by ammaratef45** [comment]
@Bahaaib Danke :+1:

**Comment by ammaratef45** [comment]
@rultor merge

**Comment by 0crat** [comment]
Job `gh:ammaratef45/attendance#35` is not assigned, can't get performer

**Comment by 0crat** [comment]
The job [#35](https://github.com/ammaratef45/attendance/issues/35) is now out of scope

**Pull request merged**